File: Quera/Easy/6404/6404_2.cpp
# https://quera.org/problemset/6404


#include <iostream>
#include <vector>
#include <set>
#include <unordered_set>
#include <algorithm>

using namespace std;

class Reaction {
public:
    vector<int> pre_materials;
    vector<int> response_materials;

    Reaction(const vector<int>& pre, const vector<int>& response) {
        pre_materials = pre;
        response_materials = response;
    }
};

int main() {
    int n, m, k;
    cin >> n >> m >> k;

    unordered_set<int> owned_materials;
    for (int i = 0; i < m; ++i) {
        int material;
        cin >> material;
        owned_materials.insert(material);
    }

    vector<Reaction> reactions;
    reactions.reserve(k); // Preallocate space to avoid multiple reallocations

    for (int i = 0; i < k; ++i) {
        int pre_n, response_n;
        cin >> pre_n >> response_n;

        vector<int> pre_materials(pre_n);
        for (int j = 0; j < pre_n; ++j) {
            cin >> pre_materials[j];
        }

        vector<int> response_materials(response_n);
        for (int j = 0; j < response_n; ++j) {
            cin >> response_materials[j];
        }

        reactions.emplace_back(pre_materials, response_materials);
    }

    bool progress = true;
    while (progress) {
        progress = false;
        for (size_t i = 0; i < reactions.size(); ++i) {
            bool can_react = true;
            for (int pre_mat : reactions[i].pre_materials) {
                if (owned_materials.find(pre_mat) == owned_materials.end()) {
                    can_react = false;
                    break;
                }
            }

            if (can_react) {
                owned_materials.insert(reactions[i].response_materials.begin(), reactions[i].response_materials.end());
                reactions.erase(reactions.begin() + i);
                progress = true;
                break;
            }
        }
    }

    // Print results
    cout << owned_materials.size() << endl;
    vector<int> sorted_owned_materials(owned_materials.begin(), owned_materials.end());
    sort(sorted_owned_materials.begin(), sorted_owned_materials.end());

    for (int material : sorted_owned_materials) {
        cout << material << " ";
    }
    cout << endl;

    return 0;
}

File: Quera/Easy/6404/6404_2.py
# https://quera.org/problemset/6404

class Reaction:

    def __init__(self, pre_materials: set, response_materials: set):
        self.pre_materials = pre_materials
        self.response_materials = response_materials


reactions = []

n, m, k = map(int, input().split())
owned_materials = {a: False for a in range(1, n+1)}
for i in map(int, input().split()):
    owned_materials[i] = True

for i in range(k):
    pre_n, response_n = map(int, input().split())
    pre_materials = set(map(int, input().split()))
    response_materials = set(map(int, input().split()))

    reactions.append(Reaction(pre_materials, response_materials))


while True:

    for i in range(len(reactions)):
        pre_materials_i = reactions[i].pre_materials
        if all(((owned_materials[pre_mat]) for pre_mat in pre_materials_i)):
            for response_mat in reactions[i].response_materials:
                owned_materials[response_mat] = True
            reactions.pop(i)
            break
    else:
        break


# print results
owned_materials = [key for key in owned_materials.keys() if owned_materials[key]]
print(len(owned_materials))
owned_materials = sorted(owned_materials)
print(*owned_materials, end=' ')
